crates/walrus-service/src/node.rs

```diff
@@ -2717,6 +2677,195 @@ impl StorageNodeInner {
+            }),
+            Err(e) => {
+                if e.is_panic() {
+                    std::panic::resume_unwind(e.into_panic());
+                }
+                Err(e)
+                    .context("blocking thread failed to retrieve sliver")
+                    .map_err(RetrieveSliverError::Internal)
+            }
+        }
+    }
+
+    /// Retrieve the recovery symbol without calling `validate_blob_access`.
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn retrieve_recovery_symbol_unchecked(
+        &self,
+        blob_id: &BlobId,
+        symbol_id: SymbolId,
+        sliver_type: Option<SliverType>,
+        encoding_type: EncodingType,
+        worker: RecoverySymbolService,
+    ) -> Result<GeneralRecoverySymbol, RetrieveSymbolError> {
+        let n_shards = self.n_shards();
+
+        let primary_index = symbol_id.primary_sliver_index();
+        self.check_index(primary_index)?;
+        let primary_pair_index = primary_index.to_pair_index::<Primary>(n_shards);
+
+        let secondary_index = symbol_id.secondary_sliver_index();
+        self.check_index(secondary_index)?;
+        let secondary_pair_index = secondary_index.to_pair_index::<Secondary>(n_shards);
+
+        let owned_shards = self.owned_shards_at_latest_epoch();
+
+        // In the event that neither of the slivers are assigned to this shard use this error,
+        // otherwise it is overwritten.
+        let mut final_error = RetrieveSymbolError::SymbolNotPresentAtShards;
+
+        let mut worker = Some(worker);
+
+        for target_sliver_type in [SliverType::Secondary, SliverType::Primary]
+            .into_iter()
+            .filter(|target_type| sliver_type.is_none() || sliver_type == Some(*target_type))
+        {
+            let (source_pair_index, target_pair_index) = match target_sliver_type {
+                Axis::Primary => (secondary_pair_index, primary_pair_index),
+                Axis::Secondary => (primary_pair_index, secondary_pair_index),
+            };
+
+            let required_shard = &source_pair_index.to_shard_index(n_shards, blob_id);
+            if !owned_shards.contains(required_shard) {
+                // This node does not manage the shard owning the source pair.
+                continue;
+            }
+
+            // Since the caller was granted a ready-worker, we ensure that both attempts are made,
+            // and avoid failing the second attempt with Unavailable. Note that in most cases,
+            // the node has only a single matching shard and so this loop and code only runs once.
+            let wait_for_worker = async {
+                match worker.take() {
+                    Some(worker) => worker,
+                    None => self.wait_for_ready_symbol_service().await,
+                }
+            };
+
+            let (mut worker, sliver_result) = tokio::join!(
+                wait_for_worker,
+                self.retrieve_sliver_unchecked(
```

Comment:
nit: should this be part of the RecoverySymbolService so that the overhead of reading DB is counted inside of serving the request?